File: app/src/main/java/com/acdmt/render/StimulusRenderer.kt
package com.acdmt.render

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.RectF
import android.util.AttributeSet
import android.view.View
import com.acdmt.model.Hemifield
import com.acdmt.model.StimulusItem
import com.acdmt.model.TrialPhase
import com.acdmt.model.TrialPlan
import kotlin.math.min

class StimulusRenderer @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null
) : View(context, attrs) {

    private var phase: TrialPhase = TrialPhase.IDLE
    private var currentTrial: TrialPlan? = null
    private var cueHemifield: Hemifield? = null
    private var externalScreenWidth = 0
    private var externalScreenHeight = 0

    private val textPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        color = Color.WHITE
        textAlign = Paint.Align.CENTER
    }

    private val stimulusPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        style = Paint.Style.FILL
    }

    private val stimulusBorderPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        color = Color.argb(190, 0, 0, 0)
        style = Paint.Style.STROKE
        strokeWidth = 2f
    }

    init {
        setBackgroundColor(BACKGROUND_COLOR)
        isFocusable = true
    }

    fun setExternalScreenMetrics(widthPixels: Int, heightPixels: Int) {
        externalScreenWidth = widthPixels
        externalScreenHeight = heightPixels
        invalidate()
    }

    fun showBlank(blankPhase: TrialPhase = TrialPhase.ITI) {
        phase = blankPhase
        currentTrial = null
        cueHemifield = null
        invalidate()
    }

    fun showFixation(fixationPhase: TrialPhase = TrialPhase.FIXATION) {
        phase = fixationPhase
        currentTrial = null
        cueHemifield = null
        invalidate()
    }

    fun showCue(hemifield: Hemifield) {
        phase = TrialPhase.CUE
        currentTrial = null
        cueHemifield = hemifield
        invalidate()
    }

    fun showEncoding(trial: TrialPlan) {
        phase = TrialPhase.ENCODING
        currentTrial = trial
        cueHemifield = trial.cuedHemifield
        invalidate()
    }

    fun showMaintenance() {
        phase = TrialPhase.MAINTENANCE
        invalidate()
    }

    fun showRetrieval(trial: TrialPlan) {
        phase = TrialPhase.RETRIEVAL
        currentTrial = trial
        cueHemifield = trial.cuedHemifield
        invalidate()
    }

    fun showFinished() {
        phase = TrialPhase.FINISHED
        currentTrial = null
        cueHemifield = null
        invalidate()
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        canvas.drawColor(BACKGROUND_COLOR)

        val geometry = calculateGeometry()
        when (phase) {
            TrialPhase.FIXATION,
            TrialPhase.MAINTENANCE -> drawFixation(canvas, geometry)
            TrialPhase.CUE -> drawCue(canvas, geometry)
            TrialPhase.ENCODING -> {
                drawFixation(canvas, geometry)
                currentTrial?.let { drawStimuli(canvas, it.memoryItems, geometry) }
            }
            TrialPhase.RETRIEVAL -> {
                drawFixation(canvas, geometry)
                currentTrial?.let { drawStimuli(canvas, it.testItems, geometry) }
            }
            TrialPhase.IDLE,
            TrialPhase.ITI,
            TrialPhase.FINISHED -> Unit
        }
    }

    private fun calculateGeometry(): Geometry {
        val screenWidth = when {
            width > 0 -> width
            externalScreenWidth > 0 -> externalScreenWidth
            else -> resources.displayMetrics.widthPixels
        }.toFloat()

        val screenHeight = when {
            height > 0 -> height
            externalScreenHeight > 0 -> externalScreenHeight
            else -> resources.displayMetrics.heightPixels
        }.toFloat()

        val fixationX = screenWidth * 0.5f
        val fixationY = screenHeight * 0.5f
        val sideMargin = screenWidth * 0.045f
        val centralGap = screenWidth * 0.105f
        val verticalTop = screenHeight * 0.20f
        val verticalBottom = screenHeight * 0.74f
        val squareSide = min(screenWidth * 0.068f, screenHeight * 0.058f)

        val leftHemifield = RectF(
            sideMargin,
            verticalTop,
            fixationX - centralGap,
            verticalBottom
        )
        val rightHemifield = RectF(
            fixationX + centralGap,
            verticalTop,
            screenWidth - sideMargin,
            verticalBottom
        )
        val leftStimulusBox = centeredBoxIn(leftHemifield, widthScale = 0.96f, heightScale = 0.72f)
        val rightStimulusBox = centeredBoxIn(rightHemifield, widthScale = 0.96f, heightScale = 0.72f)

        return Geometry(
            screenWidth = screenWidth,
            screenHeight = screenHeight,
            fixationX = fixationX,
            fixationY = fixationY,
            squareSide = squareSide,
            leftStimulusBox = leftStimulusBox,
            rightStimulusBox = rightStimulusBox
        )
    }

    private fun drawFixation(canvas: Canvas, geometry: Geometry) {
        textPaint.color = Color.WHITE
        textPaint.textSize = min(geometry.screenWidth, geometry.screenHeight) * 0.11f
        drawCenteredText(canvas, "+", geometry.fixationX, geometry.fixationY, textPaint)
    }

    private fun drawCue(canvas: Canvas, geometry: Geometry) {
        val symbol = cueHemifield?.cueSymbol().orEmpty()
        textPaint.color = Color.WHITE
        textPaint.textSize = min(geometry.screenWidth, geometry.screenHeight) * 0.16f
        drawCenteredText(canvas, symbol, geometry.fixationX, geometry.fixationY, textPaint)
    }

    private fun drawStimuli(canvas: Canvas, items: List<StimulusItem>, geometry: Geometry) {
        items.forEach { item ->
            val rect = rectFor(item, geometry)
            stimulusPaint.color = item.color
            canvas.drawRect(rect, stimulusPaint)
            stimulusBorderPaint.strokeWidth = geometry.squareSide * 0.045f
            canvas.drawRect(rect, stimulusBorderPaint)
        }
    }

    private fun rectFor(item: StimulusItem, geometry: Geometry): RectF {
        val box = if (item.hemifield == Hemifield.LEFT) {
            geometry.leftStimulusBox
        } else {
            geometry.rightStimulusBox
        }
        val centerX = box.centerX()
        val centerY = box.centerY()
        val horizontalRadius = ((box.width() - geometry.squareSide) * 0.5f).coerceAtLeast(0f)
        val verticalRadius = ((box.height() - geometry.squareSide) * 0.5f).coerceAtLeast(0f)
        val halfSide = geometry.squareSide * 0.5f

        val squareCenterX = centerX + item.slot.xFraction * horizontalRadius
        val squareCenterY = centerY + item.slot.yFraction * verticalRadius

        val left = (squareCenterX - halfSide).coerceIn(box.left, box.right - geometry.squareSide)
        val top = (squareCenterY - halfSide).coerceIn(box.top, box.bottom - geometry.squareSide)
        return RectF(left, top, left + geometry.squareSide, top + geometry.squareSide)
    }

    private fun centeredBoxIn(bounds: RectF, widthScale: Float, heightScale: Float): RectF {
        val boxWidth = bounds.width() * widthScale
        val boxHeight = bounds.height() * heightScale
        val left = bounds.centerX() - boxWidth * 0.5f
        val top = bounds.centerY() - boxHeight * 0.5f
        return RectF(left, top, left + boxWidth, top + boxHeight)
    }

    private fun drawCenteredText(
        canvas: Canvas,
        text: String,
        centerX: Float,
        centerY: Float,
        paint: Paint
    ) {
        val metrics = paint.fontMetrics
        val baseline = centerY - (metrics.ascent + metrics.descent) * 0.5f
        canvas.drawText(text, centerX, baseline, paint)
    }

    private data class Geometry(
        val screenWidth: Float,
        val screenHeight: Float,
        val fixationX: Float,
        val fixationY: Float,
        val squareSide: Float,
        val leftStimulusBox: RectF,
        val rightStimulusBox: RectF
    )

    companion object {
        val BACKGROUND_COLOR: Int = Color.rgb(45, 45, 48)
    }
}

File: adaptive_wmapp_flutter/lib/widgets/stimulus_renderer.dart
import 'dart:math';
import 'package:flutter/material.dart';
import '../models/experiment_models.dart';

class StimulusRenderer extends StatelessWidget {
  final TrialPhase phase;
  final TrialPlan? currentTrial;
  final Hemifield? cueHemifield;

  const StimulusRenderer({
    super.key,
    required this.phase,
    this.currentTrial,
    this.cueHemifield,
  });

  static const Color backgroundColor = Color.fromARGB(255, 45, 45, 48);

  @override
  Widget build(BuildContext context) {
    return Container(
      color: backgroundColor,
      width: double.infinity,
      height: double.infinity,
      child: CustomPaint(
        painter: _StimulusPainter(
          phase: phase,
          currentTrial: currentTrial,
          cueHemifield: cueHemifield,
        ),
      ),
    );
  }
}

class _StimulusPainter extends CustomPainter {
  final TrialPhase phase;
  final TrialPlan? currentTrial;
  final Hemifield? cueHemifield;

  _StimulusPainter({
    required this.phase,
    this.currentTrial,
    this.cueHemifield,
  });

  @override
  void paint(Canvas canvas, Size size) {
    final geometry = _calculateGeometry(size);

    switch (phase) {
      case TrialPhase.fixation:
      case TrialPhase.maintenance:
        _drawFixation(canvas, geometry);
        break;
      case TrialPhase.cue:
        _drawCue(canvas, geometry);
        break;
      case TrialPhase.encoding:
        _drawFixation(canvas, geometry);
        if (currentTrial != null) {
          _drawStimuli(canvas, currentTrial!.memoryItems, geometry);
        }
        break;
      case TrialPhase.retrieval:
        _drawFixation(canvas, geometry);
        if (currentTrial != null) {
          _drawStimuli(canvas, currentTrial!.testItems, geometry);
        }
        break;
      case TrialPhase.idle:
      case TrialPhase.iti:
      case TrialPhase.finished:
        break;
    }
  }

  _Geometry _calculateGeometry(Size size) {
    final screenWidth = size.width;
    final screenHeight = size.height;

    final fixationX = screenWidth * 0.5;
    final fixationY = screenHeight * 0.5;
    final sideMargin = screenWidth * 0.045;
    final centralGap = screenWidth * 0.105;
    final verticalTop = screenHeight * 0.20;
    final verticalBottom = screenHeight * 0.74;
    final squareSide = min(screenWidth * 0.068, screenHeight * 0.058);

    final leftHemifield = Rect.fromLTRB(
      sideMargin,
      verticalTop,
      fixationX - centralGap,
      verticalBottom,
    );
    final rightHemifield = Rect.fromLTRB(
      fixationX + centralGap,
      verticalTop,
      screenWidth - sideMargin,
      verticalBottom,
    );

    final leftStimulusBox = _centeredBoxIn(leftHemifield, 0.96, 0.72);
    final rightStimulusBox = _centeredBoxIn(rightHemifield, 0.96, 0.72);

    return _Geometry(
      screenWidth: screenWidth,
      screenHeight: screenHeight,
      fixationX: fixationX,
      fixationY: fixationY,
      squareSide: squareSide,
      leftStimulusBox: leftStimulusBox,
      rightStimulusBox: rightStimulusBox,
    );
  }

  Rect _centeredBoxIn(Rect bounds, double widthScale, double heightScale) {
    final boxWidth = bounds.width * widthScale;
    final boxHeight = bounds.height * heightScale;
    final left = bounds.center.dx - boxWidth * 0.5;
    final top = bounds.center.dy - boxHeight * 0.5;
    return Rect.fromLTWH(left, top, boxWidth, boxHeight);
  }

  void _drawFixation(Canvas canvas, _Geometry geometry) {
    final textSize = min(geometry.screenWidth, geometry.screenHeight) * 0.11;
    _drawCenteredText(canvas, '+', geometry.fixationX, geometry.fixationY, textSize);
  }

  void _drawCue(Canvas canvas, _Geometry geometry) {
    final symbol = cueHemifield?.cueSymbol() ?? '';
    final textSize = min(geometry.screenWidth, geometry.screenHeight) * 0.16;
    _drawCenteredText(canvas, symbol, geometry.fixationX, geometry.fixationY, textSize);
  }

  void _drawStimuli(Canvas canvas, List<StimulusItem> items, _Geometry geometry) {
    final borderPaint = Paint()
      ..color = const Color.fromARGB(190, 0, 0, 0)
      ..style = PaintingStyle.stroke
      ..strokeWidth = geometry.squareSide * 0.045;

    for (var item in items) {
      final rect = _rectFor(item, geometry);
      final fillPaint = Paint()
        ..color = item.color
        ..style = PaintingStyle.fill;
      canvas.drawRect(rect, fillPaint);
      canvas.drawRect(rect, borderPaint);
    }
  }

  Rect _rectFor(StimulusItem item, _Geometry geometry) {
    final box = item.hemifield == Hemifield.left
        ? geometry.leftStimulusBox
        : geometry.rightStimulusBox;

    final centerX = box.center.dx;
    final centerY = box.center.dy;
    final horizontalRadius = max(0.0, (box.width - geometry.squareSide) * 0.5);
    final verticalRadius = max(0.0, (box.height - geometry.squareSide) * 0.5);
    final halfSide = geometry.squareSide * 0.5;

    final squareCenterX = centerX + item.slot.xFraction * horizontalRadius;
    final squareCenterY = centerY + item.slot.yFraction * verticalRadius;

    final left = (squareCenterX - halfSide).clamp(box.left, box.right - geometry.squareSide);
    final top = (squareCenterY - halfSide).clamp(box.top, box.bottom - geometry.squareSide);
    return Rect.fromLTWH(left, top, geometry.squareSide, geometry.squareSide);
  }

  void _drawCenteredText(Canvas canvas, String text, double centerX, double centerY, double fontSize) {
    final textPainter = TextPainter(
      text: TextSpan(
        text: text,
        style: TextStyle(color: Colors.white, fontSize: fontSize, height: 1.0),
      ),
      textDirection: TextDirection.ltr,
    );
    textPainter.layout();
    final offset = Offset(centerX - textPainter.width * 0.5, centerY - textPainter.height * 0.5);
    textPainter.paint(canvas, offset);
  }

  @override
  bool shouldRepaint(covariant _StimulusPainter oldDelegate) {
    return phase != oldDelegate.phase ||
        currentTrial != oldDelegate.currentTrial ||
        cueHemifield != oldDelegate.cueHemifield;
  }
}

class _Geometry {
  final double screenWidth;
  final double screenHeight;
  final double fixationX;
  final double fixationY;
  final double squareSide;
  final Rect leftStimulusBox;
  final Rect rightStimulusBox;

  _Geometry({
    required this.screenWidth,
    required this.screenHeight,
    required this.fixationX,
    required this.fixationY,
    required this.squareSide,
    required this.leftStimulusBox,
    required this.rightStimulusBox,
  });
}
